### PR TITLE
ci: coccinelle: Add checker for rules 5.7, 14.4 and 21.2

### PR DESCRIPTION
**Pull request by ceolin**
This can spot most violations recently fixed in 
https://github.com/zephyrproject-rtos/zephyr/pull/33569
https://github.com/zephyrproject-rtos/zephyr/pull/33562

### CODE REVIEW

File: scripts/ci/guideline_check.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        description="Check if change requires full twister")
+    parser.add_argument('-c', '--commits', default=None,
+                        help="Commit range in the form: a..b")
+    return parser.parse_args()
+
+
+def main():
+    args = parse_args()
+    if not args.commits:
+        exit(1)
```
Comment (nashif) [inline scripts/ci/guideline_check.py:50]: exit("missing commit range")

File: scripts/ci/guideline_check.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+                    violation = "{}:{}".format(f.path, line.target_line_no)
+                    if violation in violations:
+                        print(
+                            "{}:{}".format(
+                                violation, "\t\n".join(
+                                    violations[violation])))
+
+
+if __name__ == "__main__":
+    main()
```
Comment (nashif) [inline scripts/ci/guideline_check.py:83]: main() should return something else other 0 if the scan fails and violations are found.

File: scripts/ci/guideline_check.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                f.path,
+                **sh_special_args)
+            parse_coccinelle(cocci, violations)
+
+        for hunk in f:
+            for line in hunk:
+                if line.is_added:
+                    violation = "{}:{}".format(f.path, line.target_line_no)
+                    if violation in violations:
+                        print(
```
Comment (nashif) [inline scripts/ci/guideline_check.py:76]: we should add an option to send the output to a file.

**Review by nashif — CHANGES_REQUESTED** [review]
see https://github.com/nashif/zephyr/commit/5165f48cca4318f348cccd9307c6d1df2bb541e0 and  https://github.com/nashif/zephyr/commit/1ae7e44963bc3efce0d97dd5d883ce4f807ce351

**Review by nashif — APPROVED** [review]

File: scripts/ci/guideline_check.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+                                fp.write("{}:{}\n".format(
+                                    violation, "\t\n".join(
+                                        violations[violation])))
+                        else:
+                            print(
+                                "{}:{}".format(
+                                    violation, "\t\n".join(
+                                        violations[violation])))
+
+        return numViolations
```
Comment (nashif) [inline scripts/ci/guideline_check.py:93]: indent here is wrong

**Review by nashif — CHANGES_REQUESTED** [review]

File: scripts/ci/guideline_check.py (hunk 4 above)
Comment (ceolin) [reply scripts/ci/guideline_check.py]: arghh, good catch

**Review by nashif — APPROVED** [review]

File: scripts/ci/guideline_check.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    return parser.parse_args()
+
+
+def main():
+    args = parse_args()
+    if not args.commits:
+        exit("missing commit range")
+
+    # pylint does not like the 'sh' library
+    # pylint: disable=too-many-function-args,unexpected-keyword-arg
```
Comment (jenmwms) [inline scripts/ci/guideline_check.py:56]: Is this command for pylint disable left over from debug perhaps? Or is this a suggestion in special scenarios?

Comment (ceolin) [reply scripts/ci/guideline_check.py:56]: nope, this is due a pylint issue.

Comment (jenmwms) [reply scripts/ci/guideline_check.py:56]: I see the commit adding this says "suppress pylint flase positive on 'sh' library". Does this commented line execute?

**Review by jenmwms — APPROVED** [review]

**Pull request merged**